### PR TITLE
feat(addressbook): Contact address must be valid to be added

### LINKED ISSUE #1013: [FEATURE REQUEST]: Address verification when adding an address in the addressbook
**Is your feature request related to a problem? Please describe.**
When adding an address to a contact in the addressbook, you can enter anything in the address field it will be valid.

Capture d’écran 2021-06-22 à 16 03 39

### PR DESCRIPTION
**Pull request by SylEze**
Closes #1013

### CODE REVIEW

**Comment by smk762** [comment]
We could improve this error message 
image
Error also appears while inputting a valid address, until final char is input. Might be better to not show error until clicking "validate" if not valid. "Validate" button implies it is triggering validation, so showing errors earlier seems premature.

Select wallet type filter seems to display all the generic wallet types regardless of input
image

Actual validation looks to be working for BEP20/ERC20/Smartchain and a handful of utxo coins I tested.

Lowercase ETH addresses show checksum error, but the error message is cut short by overflow, and there is not "convert" button to change it to mixed case
image

**Comment by smk762** [comment]
I seem to be getting `Invalid address` regardless of what I enter, even a correct address as copied from wallet page with receive button.

**Comment by SylEze** [comment]
> Select wallet type filter seems to display all the generic wallet types regardless of input
> image
> 

I think this should be put in a separated issue, will work on it ASAP

**Comment by smk762** [comment]
Still some wierdness here. Entered addresses disappearing. 
Might be related to 
```
[19:41:12] [warning] [main.prerequisites.hpp:92] [891725]: file:///home/smk762/Downloads/atomicdex-desktop-ubuntu-qt-5-15-2-4da31af.zip/atomicdex-desktop-linux-4da31af/AntaraAtomicDexAppDir/usr/qml/QtQuick/Controls/ScrollView.qml:193:5: Unable to assign TableViewStyle_QMLTYPE_1775 to Style_QMLTYPE_334
[19:41:12] [warning] [main.prerequisites.hpp:92] [891725]: file:///home/smk762/Downloads/atomicdex-desktop-ubuntu-qt-5-15-2-4da31af.zip/atomicdex-desktop-linux-4da31af/AntaraAtomicDexAppDir/usr/qml/QtQuick/Controls/Styles/Desktop/TableViewStyle.qml:60: TypeError: Cannot read property 'activeFocus' of null
[19:41:12] [warning] [main.prerequisites.hpp:92] [891725]: file:///home/smk762/Downloads/atomicdex-desktop-ubuntu-qt-5-15-2-4da31af.zip/atomicdex-desktop-linux-4da31af/AntaraAtomicDexAppDir/usr/qml/QtQuick/Controls/Styles/Desktop/TableViewStyle.qml:48:5: Unable to assign TableView_QMLTYPE_425 to BasicTableView_QMLTYPE_1752
[19:41:12] [warning] [main.prerequisites.hpp:92] [891725]: file:///home/smk762/Downloads/atomicdex-desktop-ubuntu-qt-5-15-2-4da31af.zip/atomicdex-desktop-linux-4da31af/AntaraAtomicDexAppDir/usr/qml/QtQuick/Controls/Styles/Desktop/TableViewStyle.qml:52: TypeError: Cannot read property 'enabled' of null
```
Also allowing KMD address for ERC20 / BEP20 / QRC
image

**Comment by SylEze** [comment]
> Still some wierdness here. Entered addresses disappearing again after login/logout
> Might be related to
> 
> ```
> [19:41:12] [warning] [main.prerequisites.hpp:92] [891725]: file:///home/smk762/Downloads/atomicdex-desktop-ubuntu-qt-5-15-2-4da31af.zip/atomicdex-desktop-linux-4da31af/AntaraAtomicDexAppDir/usr/qml/QtQuick/Controls/ScrollView.qml:193:5: Unable to assign TableViewStyle_QMLTYPE_1775 to Style_QMLTYPE_334
> [19:41:12] [warning] [main.prerequisites.hpp:92] [891725]: file:///home/smk762/Downloads/atomicdex-desktop-ubuntu-qt-5-15-2-4da31af.zip/atomicdex-desktop-linux-4da31af/AntaraAtomicDexAppDir/usr/qml/QtQuick/Controls/Styles/Desktop/TableViewStyle.qml:60: TypeError: Cannot read property 'activeFocus' of null
> [19:41:12] [warning] [main.prerequisites.hpp:92] [891725]: file:///home/smk762/Downloads/atomicdex-desktop-ubuntu-qt-5-15-2-4da31af.zip/atomicdex-desktop-linux-4da31af/AntaraAtomicDexAppDir/usr/qml/QtQuick/Controls/Styles/Desktop/TableViewStyle.qml:48:5: Unable to assign TableView_QMLTYPE_425 to BasicTableView_QMLTYPE_1752
> [19:41:12] [warning] [main.prerequisites.hpp:92] [891725]: file:///home/smk762/Downloads/atomicdex-desktop-ubuntu-qt-5-15-2-4da31af.zip/atomicdex-desktop-linux-4da31af/AntaraAtomicDexAppDir/usr/qml/QtQuick/Controls/Styles/Desktop/TableViewStyle.qml:52: TypeError: Cannot read property 'enabled' of null
> ```
> 
> Also allowing KMD address for ERC20 / BEP20 / QRC
> image

This warning also exists in `dev` branch, did you click on `Confirm` before logout ?

**Comment by smk762** [comment]
Just saw a new one
```
[20:16:01] [debug] [qt.wallet.page.cpp:738] [896438]: resp validateaddress: [{"result":{"is_valid":true}}]
[20:16:01] [debug] [main.prerequisites.hpp:86] [896436]: AddressBook: Address QRC-20:test:RJNTCts6afBAEMi839GHGG6ePTovwodmKz created for contact test
[20:16:01] [warning] [main.prerequisites.hpp:92] [896436]: file:///home/smk762/Downloads/atomicdex-desktop-ubuntu-qt-5-15-2-4da31af.zip/atomicdex-desktop-linux-4da31af/AntaraAtomicDexAppDir/usr/qml/QtQuick/Controls/Private/BasicTableView.qml:571: TypeError: Cannot read property 'textColor' of null

```

**Comment by SylEze** [comment]
Should also close #1123 now

File: atomic_defi_design/assets/languages/atomic_defi_en.ts [hunk 1]
```diff
@@ -2,19 +2,6 @@
...
-        <translation>Yes</translation>
-    </message>
-    <message>
-        <source>&amp;No</source>
-        <translation>No</translation>
-    </message>
-    <message>
-        <source>Cancel</source>
-        <translation>Cancel</translation>
-    </message>
```
Comment (Milerius) [inline atomic_defi_design/assets/languages/atomic_defi_en.ts:17]: please sync with dev first, you delete an important change here.

File: atomic_defi_design/assets/languages/atomic_defi_ru.ts [hunk 2]
```diff
@@ -2,19 +2,6 @@
...
-        <translation>Yes</translation>
-    </message>
-    <message>
-        <source>&amp;No</source>
-        <translation>No</translation>
-    </message>
-    <message>
-        <source>Cancel</source>
-        <translation>Cancel</translation>
-    </message>
```
Comment (Milerius) [inline atomic_defi_design/assets/languages/atomic_defi_ru.ts:17]: same remarks

File: atomic_defi_design/assets/languages/atomic_defi_tr.ts [hunk 3]
```diff
@@ -2,19 +2,6 @@
...
-        <translation>Yes</translation>
-    </message>
-    <message>
-        <source>&amp;No</source>
-        <translation>No</translation>
-    </message>
-    <message>
-        <source>Cancel</source>
-        <translation>Cancel</translation>
-    </message>
```
Comment (Milerius) [inline atomic_defi_design/assets/languages/atomic_defi_tr.ts:17]: same remarks

File: atomic_defi_design/assets/languages/atomic_defi_fr.ts [hunk 4]
```diff
@@ -1,21 +1,6 @@
...
-    </message>
-    <message>
-        <source>&amp;No</source>
-        <translation>Non</translation>
-    </message>
-    <message>
-        <source>Cancel</source>
-        <translation>Annuler</translation>
-    </message>
-</context>
```
Comment (Milerius) [inline atomic_defi_design/assets/languages/atomic_defi_fr.ts:18]: same remarks please do not delete this part, it's an important fix

**Comment by smk762** [comment]
remove `utxo_common:1493]`
image

unknown error (is invalid as a bep 20)
image

Should not be invalid
image

checksum detection and conversion ok, but error message persists as visible after conversion (validates ok).
https://user-images.githubusercontent.com/35845239/126604583-ef20becf-02cc-4b5d-b248-9564ef20426a.mp4

**Comment by smk762** [comment]
```
[15:42:17] [debug] [qt.wallet.page.cpp:747] [10011]: resp validateaddress: [{"error":"rpc:173] dispatcher_legacy:155] lp_coins:1288] No such coin: Smart Chain"}]
```
image

**Comment by smk762** [comment]
re: disappearing addressbook entries:
```
[16:36:03] [warning] [addressbook.cfg.cpp:48] [15491]: Addressbook config file was invalid, use empty configuration: [json.exception.parse_error.101] parse error at line 1, column 1: syntax error while parsing value - unexpected end of input; expected '[', '{', or a literal. Content was: 
```
Looking in ~/.atomic_qt, this wallet doesnt have a file, it has a folder (which is empty)
cant create a file with same name as folder.
```
[16:48:03] [info] [global.utilities.cpp:68] [17080]: creating directory /home/smk762/.atomic_qt/addressbook/TEST
[16:48:03] [warning] [main.prerequisites.hpp:92] [17080]: QIODevice::write (QFile, "/home/smk762/.atomic_qt/addressbook/TEST"): device not open
[16:48:03] [info] [addressbook.cfg.cpp:73] [17080]: Addressbook data successfully wrote in persistent data !
```
If I delete the folder and re-enter wallet, folder is created when I try and add a new addressbook contact.


https://user-images.githubusercontent.com/35845239/126616983-1f8bd3eb-7c96-4951-b6f7-5af5e743eae9.mp4

**Comment by SylEze** [comment]
> 
> 
> re: disappearing addressbook entries:
> 
> ```
> [16:36:03] [warning] [addressbook.cfg.cpp:48] [15491]: Addressbook config file was invalid, use empty configuration: [json.exception.parse_error.101] parse error at line 1, column 1: syntax error while parsing value - unexpected end of input; expected '[', '{', or a literal. Content was: 
> ```
> 
> Looking in ~/.atomic_qt, this wallet doesnt have a file, it has a folder (which is empty)
> cant create a file with same name as folder.
> 
> ```
> [16:48:03] [info] [global.utilities.cpp:68] [17080]: creating directory /home/smk762/.atomic_qt/addressbook/TEST
> [16:48:03] [warning] [main.prerequisites.hpp:92] [17080]: QIODevice::write (QFile, "/home/smk762/.atomic_qt/addressbook/TEST"): device not open
> [16:48:03] [info] [addressbook.cfg.cpp:73] [17080]: Addressbook data successfully wrote in persistent data !
> ```
> 
> If I delete the folder and re-enter wallet, folder is created when I try and add a new addressbook contact.
> adrrbook-folder.mp4

Can you try in another operating system just to be sure ? I got no problem on MacOS

**Comment by smk762** [comment]
Checksum error removed after convert :heavy_check_mark: 
Addressbook enrties persist after logout/restart :heavy_check_mark: 
Addressbook is creating file instead of folder (but not if folder exists, had to delete manually) :checkered_flag: 

Still some "unrecognised coin" errors
`[18:16:09] [debug] [qt.wallet.page.cpp:747] [25496]: resp validateaddress: [{"error":"rpc:173] dispatcher_legacy:155] lp_coins:1288] No such coin: AAVE"}]`
image

`[18:19:43] [debug] [qt.wallet.page.cpp:747] [25475]: resp validateaddress: [{"error":"rpc:173] dispatcher_legacy:155] lp_coins:1288] No such coin: AAVE"}]`
image

Still seeing error for smartchain address entry
`[18:24:01] [debug] [qt.wallet.page.cpp:747] [25499]: resp validateaddress: [{"error":"rpc:173] dispatcher_legacy:155] lp_coins:1288] No such coin: Smart Chain"}]
`

KMD-BEP20 still allows KMD R-Address
image

**Comment by smk762** [comment]
KMD-BEP20 with R-address rejected :heavy_check_mark: 
AAVE BEP20/ERC regongines and validation as expected :heavy_check_mark: 
Smartchain, BEP20, ERC parent chain working as expected :heavy_check_mark: 
Addressbook entries persisting after logout / relaunch :heavy_check_mark: 

Update error to `Invalid checksum`
image

Update error to `Unknown address validation error`
image

Instead of having a disabled "Convert button" it might be better to have that button hidden unless the mixed case error is active.

**Comment by smk762** [comment]
Thanks!
Convert button hidden unless required :heavy_check_mark: 
Error messages fixed :heavy_check_mark: 

One more issue I can see, is an error message in the send funds dialog. Not sure if this is related to address book, as it also happens without it.

https://user-images.githubusercontent.com/35845239/126761087-94650eea-29f1-4353-9fa9-fe01025f802c.mp4

https://user-images.githubusercontent.com/35845239/126761091-2fbbf1b5-9409-4ce9-be9f-1d7116d40df1.mp4

And finally, might be worth having a confirmation dialog for addressbook deletion (right not if you misclick on the trashcan accidentally it will remove the entry)

File: src/core/atomicdex/pages/qt.wallet.page.cpp [hunk 5]
```diff
@@ -336,6 +336,16 @@ namespace atomic_dex
...
+            else if (reason.contains("Invalid Checksum"))
+            {
+                reason = tr("Invalid checksum.");
+                json_result["convertible"] = false;
+            }
+            else
+            {
+                reason = tr("Unknown error.");
+                json_result["convertible"] = false;
+            }
```
Comment (Milerius) [inline src/core/atomicdex/pages/qt.wallet.page.cpp:348]: This is a faulty behavior, in every working case you will enter to this else.

**Review by smk762 — APPROVED** [review]
Sending funds dialog working :heavy_check_mark: 
Moved contact deletion to separate issue - https://github.com/KomodoPlatform/atomicDEX-Desktop/issues/1140

**Pull request merged**